collector/logs/service_test.go

```diff
@@ -43,3 +45,135 @@ func TestPipeline(t *testing.T) {
 	<-sink.DoneChan()
 	service.Close()
 }
+
+func TestLifecycle(t *testing.T) {
+	// Ensure we can open and close the service cleanly
+	// source only sends the log when it is closed, simulating a flush on close.
+	// transformer panics if told to transform when it is closed.
+	// sink panics if it receives a log after it is closed.
+	sink := sinks.NewCountingSink(1)
+	transform := &transformerpanicifclosed{}
+	transforms := []types.Transformer{transform}
+	source := newSourceSendOnClose(engine.WorkerCreator(transforms, sink))
+
+	service := &logs.Service{
+		Source:     source,
+		Transforms: transforms,
+		Sink:       sink,
+	}
+	context := context.Background()
+
+	err := service.Open(context)
+	if err != nil {
+		t.Fatalf("Failed to open service: %v", err)
+	}
+	service.Close()
+	<-sink.DoneChan()
+}
+
+type sourceSendOnClose struct {
+	workerGenerator engine.WorkerCreatorFunc
+	outputQueue     chan *types.LogBatch
+	wg              sync.WaitGroup
+
+	ctx    context.Context
+	cancel context.CancelFunc
+}
+
+func newSourceSendOnClose(workerGenerator engine.WorkerCreatorFunc) *sourceSendOnClose {
+	return &sourceSendOnClose{
+		workerGenerator: workerGenerator,
+		outputQueue:     make(chan *types.LogBatch, 1),
+	}
+}
+
+func (s *sourceSendOnClose) Open(ctx context.Context) error {
+	if s.ctx != nil {
+		panic("context already set")
+	}
+	if s.cancel != nil {
+		panic("cancel function already set")
+	}
+	s.ctx, s.cancel = context.WithCancel(ctx)
+
+	worker := s.workerGenerator("ConstSource", s.outputQueue)
+	s.wg.Add(1)
+	go func() {
+		defer s.wg.Done()
+		worker.Run()
+	}()
+	return nil
+}
+
+func (s *sourceSendOnClose) Close() error {
+	// Panic if the context is already cancelled
+	select {
+	case <-s.ctx.Done():
+		panic("context already cancelled")
+	default:
+		// Context is still active, proceed with cancellation
+	}
+
+	s.cancel()
+	batch := &types.LogBatch{}
+	batch.Ack = func() {}
+	batch.AddLiterals([]*types.LogLiteral{
+		{
+			Body: map[string]interface{}{
+				"message": "test message",
+			},
+		},
+	})
+	s.outputQueue <- batch
+	close(s.outputQueue)
+
+	s.wg.Wait()
+	return nil
+}
+
+type transformerpanicifclosed struct {
+	ctx    context.Context
+	cancel context.CancelFunc
+}
+
+func (t *transformerpanicifclosed) Open(ctx context.Context) error {
+	if t.ctx != nil {
+		panic("context already set")
+	}
+	if t.cancel != nil {
+		panic("cancel function already set")
+	}
+
+	t.ctx, t.cancel = context.WithCancel(context.Background())
+	return nil
+}
+
+func (t *transformerpanicifclosed) Close() error {
+	// Panic if the context is already cancelled
+	select {
+	case <-t.ctx.Done():
+		panic("context already cancelled")
```

Comment:
Return error vs panic?

Reply:
I think we want panics for this test-only impl to quickly assert.